src/wallet/wallet_tx_builder.cpp

```diff
@@ -226,34 +227,30 @@ ResolveNetPayment(
     // We initially resolve the payment with `MINIMUM_FEE` so that we can use the payment to
     // calculate the actual fee.
     auto initialFee = fee.value_or(MINIMUM_FEE);
-    return ValidateAmount(spendable, initialFee)
-        .and_then([&](void) {
-            return ResolvePayment(
-                    Payment(netpay.first, spendable.Total() - initialFee, netpay.second),
-                    canResolveOrchard,
-                    strategy,
-                    maxSaplingAvailable,
-                    maxOrchardAvailable,
-                    orchardOutputs)
-                .map_error([](const auto& error) -> InputSelectionError { return error; })
-                .and_then([&](const auto& rpayment) {
-                    auto finalFee = fee.value_or(CalcZIP317Fee(spendable, {rpayment}, std::nullopt));
-                    return ValidateAmount(spendable, finalFee)
-                        .and_then([&](void) {
-                            return ResolvePayment(
-                                    Payment(netpay.first, spendable.Total() - finalFee, netpay.second),
-                                    canResolveOrchard,
-                                    strategy,
-                                    maxSaplingAvailable,
-                                    maxOrchardAvailable,
-                                    orchardOutputs)
-                                .map([&](const auto& actualPayment) {
-                                    return std::make_pair(actualPayment, finalFee);
-                                })
-                                .map_error([](const auto& error) -> InputSelectionError { return error; });
-                        });
-                });
-        });
+    try_unwrap(ValidateAmount(spendable, initialFee)
+    ) or_return;
+    const ResolvedPayment rpayment = try_unwrap(ResolvePayment(
+            Payment(netpay.first, spendable.Total() - initialFee, netpay.second),
+            canResolveOrchard,
+            strategy,
+            maxSaplingAvailable,
+            maxOrchardAvailable,
+            orchardOutputs)
+    ) or_return;
```

Comment:
Notice that the `map_error` goes away because `ResolvePayment` returns a `tl::expected<ResolvedPayment, AddressResolutionError>`, and `AddressResolutionError` is a variant of `InputSelectionError`. So it just works, because there is an implicit conversion from each arm of a variant to the variant type.

Reply:
I'm not wedded to the convention I've adopted here of putting the `) or return` at the start of the line, but it does make it easier to see that there is a potential return.